### PR TITLE
Jest test has many

### PR DESCRIPTION
**Pull request by dulerong**
## Pull request type

Please check the type of change your PR introduces:
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [x] Other (please describe): test

## What is the current behaviour?

Issue Number(s): N/A

## What is the new behaviour?

- add tests for has_many.js

## Other information

N/A

### CODE REVIEW

**Comment by dulerong** [comment]
@ilunglee 
Hey, this time I squashed all my commits before pushing, could you please take a look and let me know what you think?

Thanks.

**Comment by ilunglee** [comment]
> @ilunglee Hey, this time I squashed all my commits before pushing, could you please take a look and let me know what you think?
> 
> Thanks.

@dulerong, 
When squashing your commits, please ensure you clean up the commit message. For example, I recommend changing the commit message for 9bfc4c8e8ea7b54cc79a3c5c2d208fcc24e98d8f to:
```
commit 9bfc4c8e8ea7b54cc79a3c5c2d208fcc24e98d8f
Author: dulerong <codeydu@hotmail.com>
Date:   Tue Oct 12 12:22:48 2021 +0900
    test: add tests for has_many.js
```

Please see https://github.com/CMDBrew/adminterface/blob/main/CONTRIBUTING.md#conventional-commit-messages for more information. 

Thanks 🙏

**Review by ilunglee — CHANGES_REQUESTED** [review]
Hey @dulerong, thanks for the great work! Please see my comments below.

## Commit message
When squashing your commits, please ensure you clean up the commit message. For example, I recommend changing the commit message for 9bfc4c8e8ea7b54cc79a3c5c2d208fcc24e98d8f to:
```
commit 9bfc4c8e8ea7b54cc79a3c5c2d208fcc24e98d8f
Author: dulerong <codeydu@hotmail.com>
Date:   Tue Oct 12 12:22:48 2021 +0900
    test: add tests for has_many.js
```

Please see [CONTRIBUTING.md](https://github.com/CMDBrew/adminterface/blob/main/CONTRIBUTING.md#conventional-commit-messages) for more information. 

## HTML templates for Jest testings
I'm wondering if it would make sense for us to use `jsdom` and load an example page (i.e., has_many.html) for jest testings. 
```javascript
document.body.innerHTML = fs.readFileSync("./fixtures/has_many.html");
```

It will be easier to maintain and much clearer to read this way, and we don't need to create the elements manually. See the following links for more information:
- https://freecontent.manning.com/testing-with-node-jest-and-jsdom 
- https://jestjs.io/docs/tutorial-jquery

---
## `__tests__` folder structure
Also, not related to this PR, I'm wondering if we should move the `__tests__` folder from `app/javascript/adminterface/lib/__tests__` to `app/javascript/adminterface/__tests__` folder to represent all the JavaScript files and move the files to their respective folder paths (e.g., `__tests__/has_many.spec.js` -> `__tests__/lib/has_many.spec.js`).

What do you think? Thanks!

File: app/javascript/adminterface/lib/__tests__/has_many.spec.js [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+
+  test('constructor initiates correctly', () => {
+    const spyOnBind = jest.spyOn(HasManyClass.prototype, '_bind')
+    spyOnBind.mockImplementation(() => {})
+
+    const HasMany = new HasManyClass(newNode)
+    expect(HasMany.element).toStrictEqual(newNode)
+
+    expect(HasMany.events).toStrictEqual(events)
+    expect(HasMany.options).toStrictEqual(options)
```
[the remark this answers is not in the document]
Comment (dulerong) [reply app/javascript/adminterface/lib/__tests__/has_many.spec.js:75]: @ilunglee 
You're right. Options and Events are not configurable therefore testing it wouldn't be necessary.
I'll make the amendment.

[the remark this answers is not in the document]
Comment (dulerong) [reply app/javascript/adminterface/lib/__tests__/has_many.spec.js:75]: @ilunglee 
Regarding commit message: I fixed it accordingly.

Regarding HTML templates for Jest testing: I've managed to refactor the html elements to a separate html file.

Regarding folder structure: please take a look at following stackoverflow conversation.
https://stackoverflow.com/questions/62442797/jest-folder-structure

In there, the best answer, listed 3 possible folder structure.
What you are referring to, as in having one main folder for all test files, and within this main folder having the exact same structure as the actual files being tested, is option number 3 (I believe).

I personally, would prefer option 1 or 2, either having them side by side in same folder, or having multiple __tests__ folder. It's just much easier to refer to the actual file being tested when I'm working with them.

[the remark this answers is not in the document]
Comment (ilunglee) [reply app/javascript/adminterface/lib/__tests__/has_many.spec.js:75]: @dulerong, thanks for the info. We can keep the folder structure as it is. Thanks!

[the remark this answers is not in the document]
Comment (dulerong) [reply app/javascript/adminterface/lib/__tests__/has_many.spec.js]: @ilunglee 
Hey, so I've forced push into this branch, could you please take a look?

Thanks.

**Pull request merged**